clusterloader2/pkg/imagepreload/imagepreload.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package imagepreload
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/config"
+	"k8s.io/perf-tests/clusterloader2/pkg/flags"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework/client"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util/informer"
+)
+
+const (
+	informerTimeout = time.Minute
+	manifest        = "$GOPATH/src/k8s.io/perf-tests/clusterloader2/pkg/imagepreload/manifests/daemonset.yaml"
+	namespace       = "preload"
+	pollingInterval = 5 * time.Second
+	pollingTimeout  = 5 * time.Minute
+)
+
+var images []string
+
+func init() {
+	flags.StringSliceEnvVar(&images, "node-preload-images", "NODE_PRELOAD_IMAGES", []string{}, "List of images to preload on each node in the test cluster before executing tests")
+}
+
+type controller struct {
+	config          *config.ClusterLoaderConfig
+	framework       *framework.Framework
+	templateMapping map[string]interface{}
+	images          []string
+
+	lock sync.Mutex
+}
+
+// Setup ensures every node in cluster preloads given list of images before starting tests.
+// It does it by creating a Daemonset that call "docker pull" and awaits for Node object to be updated.
+// As a side-effect of the image preloading, size of Node objects is increased.
+//
+// Preloading is skipped in kubemark or if no images have been specified.
+func Setup(conf *config.ClusterLoaderConfig, f *framework.Framework) error {
+	mapping, err := config.GetMapping(conf)
+	if err != nil {
+		return err
+	}
+
+	ctl := &controller{
+		config:          conf,
+		framework:       f,
+		templateMapping: mapping,
+		images:          images,
+	}
+	return ctl.PreloadImages()
+}
+
+func (c *controller) PreloadImages() error {
+	if len(images) == 0 {
+		klog.Warning("No images specified. Skipping image preloading")
+		return nil
+	}
+	if c.config.ClusterConfig.Provider == "kubemark" {
+		klog.Warning("Image preloading is disabled in kubemark")
+		return nil
+	}
+
+	clusterSize := c.config.ClusterConfig.Nodes
```

Comment:
Instead of relying on the config, can we use the function that we already have:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/runtimeobjects/runtimeobjects.go#L266

Reply:
I am not sure how this is supposed to work for nodes.

Reply:
I was thinking about passing the DaemonSet you are creating to this function.

Reply:
I am not sure what the benefit of doing it this way.

Reply:
The benefit is that this way it may be flake.
The first example that comes to my mind is that we are accepting the fact that some small percentage of nodes won't come up (and the cluster is still treated as healthy).
In this case, this will timeout, because if 1 node is unhealthy (or even not registered), it will never preload anything. But the number from Config still contains it.

Reply:
That sounds fine to me, that we timeout here if some nodes won't become healthy - in the name of the principle, that we should fail fast.

Reply:
My point is that if we would use what I'm suggesting, we wouldn't have to timeout, because we would know for how many pods we should wait.

Reply:
So if you don't want to do that, at the very least it requires a TODO.